packages/regl-worldview/src/commands/GLText.js

```diff
@@ -0,0 +1,393 @@
+  void main() {
+    float dist = texture2D(atlas, vTexCoord).a;
+
+    // fwidth(dist) is used to provide some anti-aliasing. However it's currently only used
+    // between outline and text, not on the outer border, because the alpha blending and
+    // depth test don't work together nicely for partially-transparent pixels.
+    if (vEnableOutline > 0.5) {
+      float screenSize = fwidth(vTexCoord.x);
+      float edgeStep = smoothstep(1.0 - cutoff - fwidth(dist), 1.0 - cutoff, dist);
+      gl_FragColor = mix(vOutlineColor, vForegroundColor, edgeStep);
+      gl_FragColor.a *= step(1.0 - outlineCutoff, dist);
+    } else {
+      gl_FragColor = vForegroundColor;
+      gl_FragColor.a *= step(1.0 - cutoff, dist);
+    }
+
+    if (gl_FragColor.a == 0.) {
+      discard;
+    }
+  }
+`;
+
+function makeTextCommand() {
+  // Keep the set of rendered characters around so we don't have to rebuild the font atlas too often.
+  const charSet = new Set();
+  const memoizedBuildAtlas = createMemoizedBuildAtlas();
+
+  const command = (regl: any) => {
+    const atlasTexture = regl.texture();
+    const drawText = regl({
+      depth: defaultDepth,
+      blend: defaultBlend,
+      primitive: "triangle strip",
+      vert,
+      frag,
+      uniforms: {
+        atlas: atlasTexture,
+        atlasSize: () => [atlasTexture.width, atlasTexture.height],
+        fontSize: FONT_SIZE,
+        cutoff: CUTOFF,
+        outlineCutoff: OUTLINE_CUTOFF,
+        srcHeight: FONT_SIZE + 2 * BUFFER,
+      },
+      instances: regl.prop("instances"),
+      count: 4,
+      attributes: {
+        position: [[0, 0], [0, -1], [1, 0], [1, -1]],
+        texCoord: [[0, 0], [0, 1], [1, 0], [1, 1]], // flipped
+        srcOffset: (ctx, props) => ({ buffer: props.srcOffsets, divisor: 1 }),
+        destOffset: (ctx, props) => ({ buffer: props.destOffsets, divisor: 1 }),
+        srcWidth: (ctx, props) => ({ buffer: props.srcWidths, divisor: 1 }),
+        scale: (ctx, props) => ({ buffer: props.scale, divisor: 1 }),
+
+        alignmentOffset: (ctx, props) => ({ buffer: props.alignmentOffset, divisor: 1 }),
+        billboard: (ctx, props) => ({ buffer: props.billboard, divisor: 1 }),
+
+        foregroundColor: (ctx, props) => ({ buffer: props.foregroundColor, divisor: 1 }),
+        outlineColor: (ctx, props) => ({ buffer: props.outlineColor, divisor: 1 }),
+        enableOutline: (ctx, props) => ({ buffer: props.enableOutline, divisor: 1 }),
+
+        posePosition: (ctx, props) => ({ buffer: props.posePosition, divisor: 1 }),
+        poseOrientation: (ctx, props) => ({ buffer: props.poseOrientation, divisor: 1 }),
+      },
+    });
+
+    return (props: $ReadOnlyArray<TextMarker & { billboard?: boolean }>) => {
+      const prevNumChars = charSet.size;
+      for (const { text } of props) {
+        for (const char of text) {
+          charSet.add(char);
+        }
+      }
+      const charsChanged = charSet.size !== prevNumChars;
```

Comment:
I like the concept of lazy generating/updating the texture only when new characters arrive from an ease of use standpoint...but it feels like there would be a lot of wins if you were able to specify the universe of characters up front instead of doing a neasted loop & set construction on ever render.  With thousands of text markers this could potentially add up to quite a bit of overhead over the course of a long scene, particularly once all the characters have been encountered.  This goes double for ros as text markers only support ASCII.  Maybe at the minimum would be nice to allow for a pre-specified universe or maybe just `ascii: true` and bypassing the set construction on every render of every instance of the command.  Nothing that needs to be added in this PR but wanted to call it out as a future perf win.

Reply:
Given that we have to iterate over each character anyway to build the attribute arrays, do you think checking whether the character is in the set really adds much extra overhead? (Chrome perf tools were giving me some trouble and not showing amount of time spent per line.)  I made sure not to actually allocate a new Set unless we found some new characters.